src/mbgl/text/glyph_atlas.cpp

```diff
@@ -87,45 +87,70 @@ GlyphAtlas::GlyphRequest& GlyphAtlas::requestRange(Entry& entry, const FontStack
     }
 
-    request.req = fileSource.request(Resource::glyphs(glyphURL, fontStack, range),
-        [this, &entry, &request, fontStack, range](Response res) {
-            if (res.error) {
-                observer->onGlyphsError(fontStack, range, std::make_exception_ptr(std::runtime_error(res.error->message)));
-            } else if (res.notModified) {
-                return;
-            } else {
-                if (!res.noContent) {
-                    std::vector<SDFGlyph> glyphs;
-
-                    try {
-                        glyphs = parseGlyphPBF(range, *res.data);
-                    } catch (...) {
-                        observer->onGlyphsError(fontStack, range, std::current_exception());
-                        return;
-                    }
-
-                    for (auto& glyph : glyphs) {
-                        entry.glyphSet.insert(std::move(glyph));
-                    }
-                }
+    request.req = fileSource.request(Resource::glyphs(glyphURL, fontStack, range), [this, fontStack, range](Response res) {
+        processResponse(res, fontStack, range);
+    });
+
+    return request;
+}
 
-                request.parsed = true;
+void GlyphAtlas::processResponse(const Response& res, const FontStack& fontStack, const GlyphRange& range) {
+    if (res.error) {
+        observer->onGlyphsError(fontStack, range, std::make_exception_ptr(std::runtime_error(res.error->message)));
+        return;
+    }
 
-                for (auto& pair : request.requestors) {
-                    GlyphRequestor& requestor = *pair.first;
-                    const std::shared_ptr<GlyphDependencies>& dependencies = pair.second;
-                    if (dependencies.unique()) {
-                        addGlyphs(requestor, *dependencies);
-                    }
-                }
+    if (res.notModified) {
+        return;
+    }
+
+    Entry& entry = entries[fontStack];
+    GlyphRequest& request = entry.ranges[range];
+
+    if (!res.noContent) {
+        std::vector<SDFGlyph> glyphs;
 
-                request.requestors.clear();
+        try {
+            glyphs = parseGlyphPBF(range, *res.data);
+        } catch (...) {
+            observer->onGlyphsError(fontStack, range, std::current_exception());
+            return;
+        }
 
-                observer->onGlyphsLoaded(fontStack, range);
+        for (auto& glyph : glyphs) {
+            auto it = entry.sdfs.find(glyph.id);
+            if (it == entry.sdfs.end()) {
+                // Glyph doesn't exist yet.
+                entry.sdfs.emplace(glyph.id, std::move(glyph));
+            } else if (it->second.metrics == glyph.metrics) {
+                if (it->second.bitmap != glyph.bitmap) {
+                    // The actual bitmap was updated; this is unsupported.
+                    Log::Warning(Event::Glyph, "Modified glyph changed bitmap represenation");
+                }
+                // At least try to update it in case it's currently unsused.
```

Comment:
This logic is ostensibly for the case where we revalidate a cached glyph PBF and receive an updated version. I'm not sure it ever happens in practice.

The two bitmaps are not necessarily equal; we might have logged the warning above and then continued through.